Fixed #34331 -- Added QuerySet.aiterator() support for prefetch_related().

john-parton opened the pull request:
I had a similar version of this code which I was using for a private project, and I noticed there was a pull request open already https://github.com/django/django/pull/16543

I'm happy to make any necessary changes.

charettes commented on tests/async/test_async_queryset.py:
```diff
@@ -48,11 +57,11 @@ async def test_aiterator(self):
-            async for m in qs:
-                pass
+        async for s in SimpleModel.objects.prefetch_related(
+            Prefetch("relatedmodel_set", to_attr="prefetched_relatedmodel")
```
Using `to_attr` is a clever way of making sure the code effectively ran [without actually using `assertNumQueries`](https://github.com/django/django/pull/16543/files#r1115127795) 🎉

charettes commented on django/db/models/query.py:
```diff
@@ -2387,6 +2409,9 @@ def prefetch_related_objects(model_instances, *related_lookups):
                 obj_list = new_obj_list
 
 
+aprefetch_related_objects = sync_to_async(prefetch_related_objects)
```
That's good, creating `sync_to_async` objects is not cheap. We started using this pattern so often that I wonder if we'll eventually want to get to a point where these objects are lazily created at the first call time like we did with `lazy_re`

I'd take the time to document this function as well since [the _sync_ version is documented](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#django.db.models.prefetch_related_objects)

john-parton replied:
That's interesting. Django has a pretty long initialization time. I would hate to continue to add to it.

I'm pushing that documentation change right now.

charettes commented:
@john-parton thanks for reviving these efforts, could you also [assign the ticket to you](https://code.djangoproject.com/ticket/34331)?

john-parton commented:
Done. I'm more active on GitHub than Django Trac, so if I go silent there, please follow up here.

I've tested the code out a fair bit, and I'm convinced it behaves more or less how you would expect.

I'm not sure that it actually performs better, but I think that's largely due to the prefetch function being synchronous. I'm but on psycopg3 yet, so I'm possibly limited there as well

felixxm commented on tests/async/test_async_queryset.py:
```diff
@@ -48,11 +57,11 @@ async def test_aiterator(self):
-        with self.assertRaisesMessage(NotSupportedError, msg):
-            async for m in qs:
-                pass
+        async for s in SimpleModel.objects.prefetch_related(
+            Prefetch("relatedmodel_set", to_attr="prefetched_relatedmodel")
+        ).aiterator():
+            count = len(s.prefetched_relatedmodel)
+            self.assertEqual(count, 1)
```
I'd assert lists of related objects, e.g.:
```suggestion
        results = []
        async for s in SimpleModel.objects.prefetch_related(
            Prefetch("relatedmodel_set", to_attr="prefetched_relatedmodel")
        ).aiterator():
            results.append(s.prefetched_relatedmodel)
        self.assertCountEqual(results, [[self.r1], [self.r2], [self.r3]])
```

john-parton replied:
That's a much better test. Updated

felixxm reviewed (commented):
@john-parton Thanks :+1: I left comments.

john-parton commented on django/db/models/query.py:
```diff
@@ -544,18 +544,40 @@ async def aiterator(self, chunk_size=2000):
+
+            return
+
+        async for item in iterable:
```
I agree. That's better. Updated.

john-parton commented:
@felixxm I made all of the requested changes. Pretty sure I got the everything updated as requested: updated test, preferring if/else over return, `versionchanged` annotation, updated release note, fixed white-space issue.

Let me know if any additional changes are required.

Thanks

john-parton commented:
It looks like my addition to the  release notes didn't work. Please advise

felixxm reviewed (commented):
@john-parton Thanks :+1: I pushed edits to release notes.

john-parton commented:
Nice! Documentation is one of my weak areas. Let me know if you need anything else. Excited to contribute again
